velox/expression/EvalCtx.h

```diff
@@ -27,8 +27,139 @@ namespace facebook::velox::exec {
+/// translate top level rows into inner rows and wrap a vector (typically
+/// generated as a result of applying an expression on the peeled vectors).
+///
+/// Typical usage pattern for peeling includes:
+/// 1. peeling a set of input vectors
+/// 2. converting relevant rows (top level rows and final selection rows)
+/// 3. Saving the current context and setting the peel
+/// 4. Applying the function or moving forward with expression eval
+/// 5. wrapping the result vector with the peel
+///
+/// Few important examples of how vectors are peeled:
+/// 1. Common dictionary layers are peeled
+///    Input Vectors: Dict1(Dict2(Flat1)), Dict1(Dict2(Const1)),
+///                   Dict1(Dict2(Dict3(Flat2)))
+///    Peeled Vectors: Flat, Const1, Dict3(Flat2)
+///    Peel: Dict1(Dict2) => collapsed into one dictionary
+///
+/// 2. Common dictionary layers are peeled
+///    Input Vectors: Dict1(Dict2(Flat1)), Dict1(Const1)),
+///                   Dict1(Dict2(Dict3(Flat2)))
+///    Peeled Vectors: Dict2(Flat), Const1, Dict2(Dict3(Flat2))
+///    Peel: Dict1
+///
+/// 3. Common dictionary layers are peeled while constant is ignored
+///    (since all valid rows translated via the common dictionary layers would
+///    point to the same constant index)
+///    Input Vectors: Dict1(Dict2(Flat1)), Const1,
+///                   Dict1(Dict2(Dict3(Flat2)))
+///    Peeled Vectors: Flat, Const1, Dict3(Flat2)
+///    Peel: Dict1(Dict2) => collapsed into one dictionary
+///
+/// 4. A single vector with constant encoding layer over a complex vector can
+///    be peeled.
+///    Input Vectors: Const1(Complex1)
+///    Peeled Vectors: Complex
+///    Peel: Const1
+///
+/// 5. All constant inputs where the peel is just a constant pointing to the
+///    first valid row. The peel itself is irrelevant but allows us to translate
+///    input rows into a single row.
+///    Input Vectors: Const1(Complex1), Const2, Const3
+///    Peeled Vectors: Const1(Complex1), Const2, Const3
+///    Peel: Const(generic constant index)
+class PeeledEncoding {
+ public:
+  /// Takes a set of vectors and peels all the common encoding layers off of
+  /// them. Returns the set of vectors obtained after peeling those layers. If
+  /// peeling is not possible then returns an empty vector.
+  std::vector<VectorPtr> getPeeledVectors(
+      std::vector<VectorPtr> vectorsToPeel,
+      const SelectivityVector& rows,
+      LocalDecodedVector& decodedVector,
+      bool canPeelsHaveNulls);
+
+  /// Utility method used to check whether an encoding is peel-able.
+  inline static bool isPeelable(VectorEncoding::Simple encoding) {
+    switch (encoding) {
+      case VectorEncoding::Simple::CONSTANT:
+      case VectorEncoding::Simple::DICTIONARY:
+      case VectorEncoding::Simple::SEQUENCE:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  /// Return the encoding of the peeled wrap.
+  /// Should only be called if getPeeledVectors() was successful.
```

Comment:
given that all these methods only make sense if `getPeeledVectors()` was successful, if feels like getPeeledVectors() should actually be the constructor. Then the remaining ones could just be static.